server/controller/src/main/java/ai/starwhale/mlops/domain/panel/PluginService.java

```diff
@@ -0,0 +1,122 @@
+import java.nio.charset.StandardCharsets;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
+
+
+@Slf4j
+@Service
+public class PluginService {
+
+    private final StorageAccessService storageAccessService;
+
+    private final StoragePathCoordinator storagePathCoordinator;
+
+    private final PanelPluginMapper panelPluginMapper;
+
+    private final PanelPluginConvertor panelPluginConvertor;
+
+    private final IdConvertor idConvertor;
+
+    private final ObjectMapper yamlMapper;
+
+    PluginService(
+            StorageAccessService storageAccessService,
+            StoragePathCoordinator storagePathCoordinator,
+            @Qualifier("yamlMapper") ObjectMapper yamlMapper,
+            PanelPluginMapper panelPluginMapper,
+            PanelPluginConvertor panelPluginConvertor,
+            IdConvertor idConvertor
+    ) {
+        this.storageAccessService = storageAccessService;
+        this.storagePathCoordinator = storagePathCoordinator;
+        this.yamlMapper = yamlMapper;
+        this.panelPluginMapper = panelPluginMapper;
+        this.panelPluginConvertor = panelPluginConvertor;
+        this.idConvertor = idConvertor;
+    }
+
+    public void installPlugin(MultipartFile multipartFile) {
+        byte[] content;
+        PanelPlugin.PluginManifest manifest;
+        // 1.save to oss
+        try (var input = multipartFile.getInputStream(); var dupIs = multipartFile.getInputStream()) {
+            content = TarFileUtil.getContentFromTarFile(dupIs, "", "manifest.json");
+            // check manifest
+            if (content == null) {
+                throw new StarwhaleApiException(new SwProcessException(SwProcessException.ErrorType.SYSTEM),
+                    HttpStatus.BAD_REQUEST);
+            }
+            manifest = yamlMapper.readValue(content, PanelPlugin.PluginManifest.class);
+
+            // check if exists
+            var plugin = panelPluginMapper.getByNameAndVersion(manifest.name, manifest.version);
+            if (plugin != null) {
+                throw new StarwhaleApiException(new SwProcessException(SwProcessException.ErrorType.DB),
+                    HttpStatus.CONFLICT);
+            }
+
+            var path = storagePathCoordinator.allocatePluginPath(manifest.name, manifest.version);
+            storageAccessService.put(path, input, multipartFile.getSize());
+        } catch (IOException e) {
+            throw new StarwhaleApiException(new SwProcessException(SwProcessException.ErrorType.STORAGE),
+                HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        // 2.update db
+        var entity = PanelPluginEntity.builder()
+                .name(manifest.name)
+                .version(manifest.version)
+                .meta(new String(content, StandardCharsets.UTF_8))
```

Comment:
saving `path`  in `PanelPluginEntity` will bring one benefit: 
- if the code inside `storagePathCoordinator.allocatePluginPath` updates, the `path` could also be found

Reply:
OK